File: DKPlugins/DKDom/DKDomDOMTokenList.js
// [MDN] https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList
// [INTERFACE] https://dom.spec.whatwg.org/#interface-domtokenlist

var DOMTokenList = function DOMTokenList() {
	
	////// Instance properties //////
	// [DOMTokenList.length](Read only) https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/length
    Object.defineProperty(this, "length", {
        get: function length() {
            return CPP_DKDomDOMTokenList_length()
        }
    })
	// [DOMTokenList.value](Read only) https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/value
    Object.defineProperty(this, "value", {
        get: function value() {
            return CPP_DKDomDOMTokenList_value()
        }
    })
	
	///// Instance methods //////
	// [DOMTokenList.item()] https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/item
    DOMTokenList.prototype.item = function item(index) {
        return CPP_DKDomDOMTokenList_item(index)
    }
	// [DOMTokenList.contains()] https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/contains
    DOMTokenList.prototype.contains = function contains(token) {
        return CPP_DKDomDOMTokenList_contains(token)
    }
	// [DOMTokenList.add()] https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/add
    DOMTokenList.prototype.add = function add(tokenN) {
        return CPP_DKDomDOMTokenList_add(tokenN)
    }
	// [DOMTokenList.remove()] https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/remove
    DOMTokenList.prototype.remove = function remove(tokenN) {
        return CPP_DKDomDOMTokenList_remove(tokenN)
    }
	// [DOMTokenList.replace()] https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/replace
    DOMTokenList.prototype.replace = function replace(oldToken, newToken) {
        return CPP_DKDomDOMTokenList_replace(oldToken, newToken)
    }
	// [DOMTokenList.supports()] https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/supports
    DOMTokenList.prototype.supports = function supports(token) {
         return CPP_DKDomDOMTokenList_supports(token)
    }
	// [DOMTokenList.toggle()] https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle
    DOMTokenList.prototype.toggle = function toggle(token, force) {
         return CPP_DKDomDOMTokenList_toggle(token, force)
    }
	// [DOMTokenList.entries()] https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/entries
    DOMTokenList.prototype.entries = function entries() {
         return CPP_DKDomDOMTokenList_entries()
    }
	// [DOMTokenList.forEach()] https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/forEach
    DOMTokenList.prototype.forEach = function forEach(callback, thisArg) {
         return CPP_DKDomDOMTokenList_forEach(callback, thisArg)
    }
	// [DOMTokenList.keys()] https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/keys
    DOMTokenList.prototype.keys = function keys() {
         return CPP_DKDomDOMTokenList_keys()
    }
	// [DOMTokenList.values()] https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/values
    DOMTokenList.prototype.values = function values() {
         return CPP_DKDomDOMTokenList_values()
    }
	
	////// toString //////
	if(this.toString() === "[object Object]"){
		this.toString = function(){
			return "[object DOMTokenList]"
		}
	}
	
    return this;
}
DOMTokenList.prototype = [];

File: DKPlugins/DKDom/DKDomDOMTokenList.cpp
/*
* This source file is part of digitalknob, the cross-platform C/C++/Javascript/Html/Css Solution
*
* For the latest information, see https://github.com/aquawicket/DigitalKnob
*
* Copyright(c) 2010 - 2023 Digitalknob Team, and contributors
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files(the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and /or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions :
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/
#include "DK/DKApp.h"
#include "DKDom/DKDom.h"
#include "DKDom/DKDomDOMTokenList.h"


bool DKDomDOMTokenList::Init(){
	DKDEBUGFUNC();

	////// Instance properties //////
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_length",	DKDomDOMTokenList::length);
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_value",	DKDomDOMTokenList::value);

	////// Instance methods //////
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_item",		DKDomDOMTokenList::item);
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_contains",	DKDomDOMTokenList::contains);
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_add",		DKDomDOMTokenList::add);
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_remove",	DKDomDOMTokenList::remove);
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_replace",	DKDomDOMTokenList::replace);
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_supports",	DKDomDOMTokenList::supports);
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_toggle",	DKDomDOMTokenList::toggle);
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_entries",	DKDomDOMTokenList::entries);
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_forEach",	DKDomDOMTokenList::forEach);
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_keys",		DKDomDOMTokenList::keys);
	DKDuktape::AttachFunction("CPP_DKDomDOMTokenList_values",	DKDomDOMTokenList::values);

	DKClass::DKCreate("DKDom/DKDomDOMTokenList.js");
	return true;
}

////// Instance properties //////
int DKDomDOMTokenList::length(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomDOMTokenList::value(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}


////// Instance methods //////
int DKDomDOMTokenList::item(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomDOMTokenList::contains(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomDOMTokenList::add(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	DKString address = duk_require_string(ctx, 0);
	Rml::Element* element = (Rml::Element*)addressToPointer(address);
	if (!element) {
		DKERROR("element invalid\n");
		duk_push_undefined(ctx);
		return true;
	}
	DKString token = duk_require_string(ctx, 1);
	element->SetPseudoClass(token, true);
	return true;
}

int DKDomDOMTokenList::remove(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	DKString address = duk_require_string(ctx, 0);
	Rml::Element* element = (Rml::Element*)addressToPointer(address);
	if (!element) {
		DKERROR("element invalid\n");
		duk_push_undefined(ctx);
		return true;
	}
	DKString token = duk_require_string(ctx, 1);
	element->SetPseudoClass(token, false);
	return true;
}

int DKDomDOMTokenList::replace(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomDOMTokenList::supports(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomDOMTokenList::toggle(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomDOMTokenList::entries(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomDOMTokenList::forEach(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomDOMTokenList::keys(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}

int DKDomDOMTokenList::values(duk_context* ctx) {
	DKDEBUGFUNC(ctx);
	return DKTODO();
}
